Implemented JDBCInputFormat for MariaDB, MySQL, Postgres and Derby. Code is covered by unit-test

emrehasan opened the pull request:
The first part (input-format) of issue #261 is implemented now and is also covered by a junit-test.

emrehasan committed:
Added JDBCInputFormat and set jdbc-mysql-dependancy to pom

supermegaciaccount committed:
Testclass/URL/getCon

Added Test Class (contains no tests yet...)
URL formatting done
Establishes Connection
Executes Query

supermegaciaccount committed:
cleanup,reachedEnd()

renamed attribute database to dbName
replaced connection establishment in configure with a call to
prepareConnection (as you planned i guess)
implemented reachedEnd() (maybe)

supermegaciaccount committed:
reachedEnd() properly implemented

previous implementation was insufficient

supermegaciaccount committed:
reachedEnd() again....

found a native method that does what we want...

supermegaciaccount committed:
started with nextRecord()

Column and their respective types are analyzed
framework for record.setField

supermegaciaccount committed:
various types are properly handled in nextRecord()

eu.stratosphere.pact.common.type.base.* allows conversion of javatypes
to values.not all types are supported, may have to create more
ourselves. (Array/BigDecimal,Date,Time,TimeStamp)

supermegaciaccount committed:
some comments added

supermegaciaccount committed:
added more types

supermegaciaccount committed:
added even more tyües

added a case for every sql type
few comments and questions in the code

supermegaciaccount committed:
added some minor tests

removed private flag from setCLassForDBType for testing

supermegaciaccount committed:
minor type changes

changed decimal/numeric to double from string
added sqlmxl

emrehasan committed:
- Added Dependency for Maven (postgres, oracle not found)
- Changed sql-type mapping a little bit and created extern method that handles the mapping
- Added new Exception that is thrown if sql-type can’t be mapped. Throwing should be passed on maybe.

emrehasan committed:
Catching new exception now

supermegaciaccount committed:
URL formatting works now, testcases added

2 testcases work, 2 fail, i dont know why :/

supermegaciaccount committed:
mariadb type check added (was mysql?)

supermegaciaccount committed:
random stuff ^^

Commit:
Preparing Apache Derby

Commit:
Merge branch 'master' of https://github.com/emrehasan/stratosphere

Conflicts:
	pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java
	pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java

Commit:
Resolved merging-conflicts

supermegaciaccount committed:
switched this.query=query and configure(parameters)

supermegaciaccount committed:
removed oracle type, mariadb testcase now successful

supermegaciaccount committed:
IT WORKS!

Changed URL formatting
resultSet and PactRecord use different column indices
Tests were changed to work with my specific own database

Commit:
Shortened tests

supermegaciaccount committed:
added Test for reachedEnd()

supermegaciaccount committed:
started with jdbcinputexample

supermegaciaccount committed:
continued with jdbcinputexample

supermegaciaccount committed:
small adjustments

created new Constructor that only takes a string
add NPE to reachedEnd() (since i got that one time and had no clue where
it happened)

supermegaciaccount committed:
Example works now

supermegaciaccount committed:
minor example adjustments

supermegaciaccount committed:
example: removed unused import

supermegaciaccount committed:
prepared two versions, waiting for confirmation

why do we have 2 setClassForDBType methods? cant we combine those?

supermegaciaccount committed:
example: prepared two versions

supermegaciaccount committed:
example: typo

supermegaciaccount committed:
test: adjustments due to InputFormat changes

supermegaciaccount committed:
format: Various changes

reordered functions
shortened switch statements
removed setClassForDBType(String) (never used)
removed nullcheck from getDBType (cant occur due to default values being
inserted)

supermegaciaccount committed:
example: remove obsolete version

supermegaciaccount committed:
test: adjustments

removed test_recached_end_mysql (works always, useless)
removed unused import
configure now called with blank configuration

supermegaciaccount committed:
format: removed obsolete version

moved string formatiing into prepareConnection
prepareConnection can now be called with a preformatted url string
added new constructor (string dbURL, string query)

supermegaciaccount committed:
format: added support for both versions

supermegaciaccount committed:
example: provided different examples

supermegaciaccount committed:
example: quick fix

supermegaciaccount committed:
example: quick fix

supermegaciaccount committed:
example: quick fix

supermegaciaccount committed:
format: changed indentation, merged branches

supermegaciaccount committed:
format: resultSet.close(),DBTypes exception

supermegaciaccount committed:
format: dbConn.close()

supermegaciaccount committed:
example: changed formatting

supermegaciaccount committed:
test: changed formatting

supermegaciaccount committed:
format: added license/javadoc/comments

supermegaciaccount committed:
test: added your code completely

supermegaciaccount committed:
test: formatting

supermegaciaccount committed:
test: comments, in-memory, removed tests

Comments added for setup
database (should) now run in-memory
removed type tests
rename blob test

emrehasan committed:
Added setclass-tests back

emrehasan committed:
Adjusted test for non available db-type

supermegaciaccount committed:
format: few adjustments

supermegaciaccount committed:
test: derby now in memory

supermegaciaccount committed:
format: whoops

protected version ended up twice in my version :>

supermegaciaccount committed:
format: DBType exception, READ COMMENT!!!

can we access the log in the test file`?

supermegaciaccount committed:
format: added InputFormat comment

supermegaciaccount committed:
format: removed NSTException

not reachable due to IAException in getDBType()

supermegaciaccount committed:
format: changed most exceptions to IAE

shit should break where it breaks, not at some nullpointer far away.
(and we can test it oh yeah!)

supermegaciaccount committed:
test: added negative type/conn/query test

supermegaciaccount committed:
test: adjusted for new exceptions

supermegaciaccount committed:
example: fixed bug

supermegaciaccount committed:
format: exception changes

emrehasan committed:
Syncing with origin project

rmetzger commented on pact/pact-common/pom.xml:
```diff
@@ -23,10 +23,25 @@
+			<groupId>org.apache.derby</groupId>
+			<artifactId>derby</artifactId>
+			<version>10.10.1.1</version>
+		</dependency>
```
I think we don't want to add all these databases as dependencies.

Does your Input format also run without the dependencies?
So that users that want to use a JDBC driver can add it for their own stratosphere job.

emrehasan replied:
Yes it should work without the dependencies but the derby-dependency is necessary to automatically build and test

rmetzger replied:
You can add dependencies with a `test` scope. Then, they are only loaded for the execution of the tests.

rmetzger commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,411 @@
+import eu.stratosphere.pact.common.type.base.PactString;
+
+/**
+ * InputFormat to read data from a database an generate PactReords.
```
typo

rmetzger commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,411 @@
+ * @see DriverManager
+ */
+public class JDBCInputFormat extends GenericInputFormat {
+
```
Concerning code style and guidelines: Could you remove the newlines after the class header, the enums etc.

rmetzger commented:
Okay, some more general feedback:
Can you make the inputFormat more generic?
It has the driver classes for some databases hardcoded. It would be nicer to allow the InputFormat to work with ALL JDBC drivers/databases.
I think the user has to supply the JDBC driver name as a string (remember the input format runs in a distributed fashion, so you need to make sure that the JDBC driver is loaded on the node that is running the input format. (So we need to test your InputFormat on a cluster) (We can do that here at the TU).

I'm not sure if it is neccessary to store all these information as class properties:

```
 +    private String dbTypeStr;
 +    private String host;
 +    private Integer port;
 +    private String dbName;
 +    private String username;
 +    private String password;
 +    private String derbyDBPath;
```

They should all be encoded in the connection string. I think you can still keep the constructor that gets a Configuration object, but it is probably better to construct the connection string there.

I'm not sure if you really need the `need_configure` variable. You should be able to determine if configuration is required based on the availability of a valid connection string (the variable name is not conform to Java variable naming!)

We don't want to have these comments

```
 /*
 +     * (non-Javadoc)
 +     * @see eu.stratosphere.pact.generic.io.InputFormat
 +     */
```

because they don't add any value (any smart IDE will point you to the right documentation).

I think it is okay to have only one example file. Write about the different variants in a comment!

Please ask me if you have any questions regarding my comments.
It is not my intention to discourage you ;) Your overall work is quite good!

rmetzger commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,239 @@
+    @BeforeClass
+    public static void setUpClass() {
+        try {
+            System.out.println("Before class");
```
remove debugging messages.

rmetzger commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,239 @@
+
+    public void configureEmbeddedDerbyConfig() {
+        jdbcInputFormat = new JDBCInputFormat("jdbc:derby:memory:ebookshop", "select * from books");
+        jdbcInputFormat.configure(null);
```
Why is it required to call the InputFormat with a null argument?
Does the user have to do this? or is it for testing purposes

zentol replied:
if the constructor is called with a string right away it doesn't matter what is passed to configure afterwards, as it is ignored. (We assume that username/password is encoded in the string)
passing null was supposed to show that.

rmetzger replied:
Okay, got it. So its a test to ensure that the configuration from the constructor is not overwritten. Then keep it.

rmetzger commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,239 @@
+    @Test
+    public void testJDBCInputFormatSettingPactRecordNormally() throws IOException {
+        PactRecord buffRecord;
+        List<PactRecord> records = new ArrayList<PactRecord>();
```
You don't need to store the PactRecords in a list!

Use only one PactRecord and pass it into the InputFormat over and over again (this is supported by the framework).
You can use a integer variable to count the number of records. Lift the checks on the fields to the while loop, then you have exactly the same validation without storing many PactRecords in a list.

rmetzger commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,411 @@
+            int column_count = rsmd.getColumnCount();
+            record.setNumFields(column_count);
+
+            for (int pos = 0; pos < column_count; pos++) {
```
If the loop starts from `pos = 1` you don't need the `pos+1` everywhere. You always add 1 to `pos`.

zentol replied:
we always add 1 to pos because resutlSet and PactRecord use different indices.
the first field of a PactRecord has index 0, the first field in a resultSet index 1.
see retrieteTypeAndFillRecord:
 record.setField(pos, new PactDouble(resultSet.getBigDecimal(pos + 1).doubleValue()));

rmetzger replied:
okay, you're right.
Seems that JDBC is a bit stupid here. Nobody starts counting from 1 ;)

zentol commented:
"Can you make the inputFormat more generic?"

if we let the user supply the driver url it should work. we could make it even more generic if we always require the user to pass the database url aswell, since that is hardcoded to the db type right now aswell.

rmetzger commented:
Okay, I would suggest to do it this way. Hardcoding database names is not a good idea, since JDBC is designed to be database independent.

You can, if you want, add the code you've already written to a static method that allows to build the database url.

Like this

```
GenericDataSource source = new GenericDataSource(new JDBCInputFormat(JDBCInputFormat.getURL("mysql", "username", "secret", "table"), "SELECT blabl"), "Data Source"); 
```

supermegaciaccount committed:
format: MINOR STUFF DONT BE MAD :(

comments changed
type fixed
formatting changed

supermegaciaccount committed:
test: MINOR STUFF AGAIN DONT BE MAD :(

changed formatting
removed debugging messages
changed testJDBCInputFormatThisNameIsTooLong()

emrehasan committed:
Added dependencies-scope so that they are only loaded in test-mode

emrehasan commented:
I think we removed comments with (non-Javadoc) prefix, but i assume my partner and me got this from your classes like you can see in GenericDataSource[85]

emrehasan committed:
- Removed enums since all jdbc-types can be used if user imports corresponding libs and sets right configuration
- Refactored methods to be more generic
- Refactored methods to be less complex (try/catch * 2 and 4 recursive if-else-cases). Should be more easier to read now
- Adjusted test-cases to be more generic and deleted unnecessary assignments and method-calls
- Configuration is not saved with all certain fields but we build the dburl and only save necessary parameters (drive name, username, password and query) now

emrehasan committed:
Deleted examples and just delivering one example with hint how to create JDBCInput that the other examples provided.

rmetzger commented:
Thanks for removing the comments.
You are right, our code is full of those comments, but we want to get rid of them: https://github.com/stratosphere/stratosphere/issues/122

supermegaciaccount committed:
format: fixed stuff (couldnt stop myself when...)

the example suddenly didnt work. (infact the code didnt even compile!)

before every commit we should make one with just formatting changes, so
we can properly compare stuff! :) sadly i remembered this to late.
Changes==============================
-constructors shouldnt call configure
-reordered functions so you dont have to scroll up and down
-most functions are now void, since they either throw an exception or
execute successfully
-moved try/catch block configure, in case we changed the underlying
functions

setClassFerDBType no longer checks for null values:
IAE exception covers that case nicely

prepareQueryExecution:
removed debugging message
removed check for setClasForDBType, since it throws an exception

preConAExe:
removed dbURL==null, clutters the code, and it will fail anyway in
preCon
if no username is provided it is assumed that u and p are encoded in the
query
since all paramaters are class fields, passing arguments is only
necessary to decide cases

I changed configure aswell, since it doesnt really make sense for all
constructors to demand url/query/driver, yet extract host/port from a
configuration(especially since we have a constructor that uses a
configuration aswell).

i would love to throw all that preQueExe stuff into a single method :(

supermegaciaccount committed:
test: fixed stuff

Commit:
Just removed import of static method which was implemented as own method.

StephanEwen commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,351 @@
+    private void retrieveTypeAndFillRecord(int pos, int type, PactRecord record) throws SQLException, NotTransformableSQLFieldException {
+        switch (type) {
+            case java.sql.Types.NULL:
+                record.setField(pos, new PactNull());
```
Save an Object by going for PactNull.INSTANCE

StephanEwen commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,351 @@
+     * @param query Query to execute.
+     */
+    public JDBCInputFormat(Configuration parameters, String query) {
+        this.driverName = parameters.getString("driver", "");
```
Can you make these Strings constants?

StephanEwen commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    private boolean isFieldNullOrEmpty(String field) {
+        return (field == null || field.equals(""));
```
fiel.length() == 0 is faster

StephanEwen commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,351 @@
+    public void configure(Configuration parameters) {
+        boolean needConfigure = isFieldNullOrEmpty(this.query) || isFieldNullOrEmpty(this.dbURL);
+        if (needConfigure) {
+            this.driverName = parameters.getString("drivername", null);
```
Use String constants please, as suggested above.

StephanEwen commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,351 @@
+        try {
+            prepareQueryExecution();
+        } catch (SQLException e) {
+            throw new IllegalArgumentException("Configure failed:\t!" + e.getMessage());
```
Please always chain exception, ie, pass the cause such that it does not get lost.

StephanEwen commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,194 @@
+            insertDataToSQLTables();
+            conn.close();
+        } catch (ClassNotFoundException except) {
+            except.printStackTrace();
```
You need to call Assert.fail() to mark the test case as failed at this position. Otherwise the test will go on and either sucecdd, or produce an error later, which one will have to track back to the setup problem.

StephanEwen commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,194 @@
+        } catch (ClassNotFoundException except) {
+            except.printStackTrace();
+        } catch (SQLException except) {
+            except.printStackTrace();
```
Same as above.

StephanEwen commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,194 @@
+        try {
+            prepareDerbyDatabase();
+        } catch (ClassNotFoundException e) {
+            e.printStackTrace();
```
Call Assert.fail() here to mark the setup as failed.

StephanEwen commented on pact/pact-common/src/test/java/eu/stratosphere/pact/common/io/JDBCInputFormatTest.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    @Test
+    public void testJDBCInputFormatSettingPactRecordNormally() throws IOException {
```
Can you check here for expected contents? The type check is good, but the contents should also match ;-)

If you make the contents a constant String[][] or Object[][], then you can have a loop that takes each row and compares it to the constant.

StephanEwen commented on pact/pact-common/src/main/java/eu/stratosphere/pact/common/io/JDBCInputFormat.java:
```diff
@@ -0,0 +1,351 @@
+    public boolean nextRecord(PactRecord record) {
+        try {
+            resultSet.next();
+            ResultSetMetaData rsmd = resultSet.getMetaData();
```
I think you can skip the three lines here. You do not need to set the number of fields, the record will adjust as you set fields.

zentol replied:
yup i checked, the record adjusts automatically, we could skip line 335.
but we still need column_count as an upper bound for the loop.

we should rename it to columnCount aswell...

StephanEwen commented:
Nice work!

I added a few comments, but otherwise it looks good to me!

rmetzger commented:
Hi guys,

once https://github.com/stratosphere/stratosphere/issues/289 is merged, can you create a new maven project in `stratosphere-addons` for the jdbc input (/output) format? Call it `stratosphere-jdbc`.

@StephanEwen: I guess this implementation of the  `JDBCInputFormat` will run with multiple input splits. Is there a way to set the DOP of the `GenericDataSource` from inside the InputFormat?
Or is it better to overwrite `createInputSplits()` ignoring the minSplits argument?
Or is a `SingleInstanceDataSource` or so required?

supermegaciaccount committed:
test: all exceptions covered, content check

shortened name of testJDBCInputFormat...
added data as constant Value[][]
added content check

shortened testUnsupportetSQLType (most of the code wouldn't be executed
due to exceptions)
added tests for nextRecord()/reachedEnd() used on non-configured formats
testclass now fails if the db cant be created

supermegaciaccount committed:
test: minor cleanup

supermegaciaccount committed:
test: removed configure methods

not used often enough to be useful (both only once)

rmetzger commented:
Hey guys,

we now merged the pull request for "stratosphere-addons" (https://github.com/stratosphere/stratosphere/tree/master/stratosphere-addons). 
Could you create a new maven project under `stratosphere-addons` called `stratosphere-jdbc`.  Move the example and JDBC Input format into the new maven project.

zentol commented:
we will create the maven project before the end of this week.

StephanEwen commented:
I assume that the data base input is a task that can not be parallelized, i.e., there is a single instance only of the data source that issues the query, and which then distributes the data among all parallel tasks of the next stage.

In order to tell the system to not parallelize an input task, you can mark the input format as unsplittable by implementing the marker interface https://github.com/stratosphere/stratosphere/blob/master/pact/pact-common/src/main/java/eu/stratosphere/pact/generic/io/UnsplittableInput.java

If that is the case for your input format, please add the UnsplittableInput interface to the list of implemented interfaces.

ankurcha commented:
Pardon my ignorance, but I think it should be possible to parallize data load on a jdbc input. The idea would be to define the correct shard key(s), which may be left for the user to specify(abstract method). 

That being said, by default we could make it a single partition or maybe based off the primary key (if any). But allowing the user to specify the column to use to split the dataset may be very powerful that can have real scalability and concurrency advantages.

The UnsplittableInput route is definitely more simple but this would also mean that if for some reason any part of the dataload fails, we would end up needing to reload the full dataset, whereas in a parallelizable case, only that partition /split would need to be retried.

fhueske commented:
You are right. That would work for parallel database systems, that can efficiently run queries on individual data partitions without having a central master involved. 
If however: 
- The database system is not running in parallel, the same machine is asked many queries from different clients. A single query that asks for all data is probably more efficient than multiple queries asking for subsets. (Seq. IO vs. random IO, query compilation, locking, etc. overhead).
- The database system is parallel but with a master coordinator, this master needs to serve all clients the results (even though it does not need to compute the result). So the master might become the bottleneck here.
  Please correct me if you think that's wrong ;-)

I think, this really depends on the database setup and the feature of the DBMS. Furthermore, the programmer needs to know the setup quite well (partition keys and ranges).
But I agree, for parallel DBMS, it would be good to have the option to shard queries over data partitions.

supermegaciaccount committed:
added stratosphere-addons

supermegaciaccount committed:
added stratosphere-addons

supermegaciaccount committed:
created stratosphere-jdbc

supermegaciaccount committed:
removed remains of addons

supermegaciaccount committed:
Revert "removed remains of addons"

This reverts commit d0224f8caeffc113db1e790a820cd50eec3dd0bd.

supermegaciaccount committed:
missed top pom.xml

supermegaciaccount committed:
example into maven, test remains removed

supermegaciaccount committed:
pact-clients dependency, rename example

supermegaciaccount committed:
removed test dependencies

supermegaciaccount committed:
moved format into maven

supermegaciaccount committed:
Revert "added stratosphere-addons"

supermegaciaccount committed:
pom

supermegaciaccount committed:
Revert "pom"

This reverts commit c8f8c79362d0e3eb99a25a0ee06aa4a92ed75ca0.

zentol commented:
Would it be possible to let the user supply multiple queries, and use each of them in a separate split?

Given a user who knows about shard keys, i feel like it would be really inconvenient for him to pass the query and keys separately, und hope that the correct query is generated.

rmetzger commented:
Regarding your question: Yes. You could overwrite the getInputSplits() method (that would for example return two splits, one for each query).
I think we should first merge the current JDBC input format into the main line. Its nicely written and covers most use-cases.

If one wants to do more advanced stuff, there is also the way to create multiple DataSources with JDBCInputFormats. Like this.

```
+---------------+
| MySQL:Shard1  |
+---------------+
                  \
                   (Union (Stratosphere))--(Further processing)--Data Sink
                  /
+---------------+
| MySQL:Shard2  |
+---------------+
```

ankurcha commented:
That seems for to me. 

> On Nov 29, 2013, at 17:06, Robert Metzger notifications@github.com wrote:
> 
> Regarding your question: Yes. You could overwrite the getInputSplits() method (that would for example return two splits, one for each query).
> I think we should first merge the current JDBC input format into the main line. Its nicely written and covers most use-cases.
> 
> If one wants to do more advanced stuff, there is also the way to create multiple DataSources with JDBCInputFormats. Like this.
> 
> +---------------+
> | MySQL:Shard1  |
> +---------------+
>                   \
>                    (Union (Stratosphere))--(Further processing)--Data Sink
>                   /
> +---------------+
> | MySQL:Shard2  |
> +---------------+
> —
> Reply to this email directly or view it on GitHub.

emrehasan committed:
- Merge branch 'master' of https://github.com/stratosphere/stratosphere
- Changed implementation of JDBCInputFormat considering @StephanEwan -s comments

Conflicts:
	stratosphere-addons/pom.xml

emrehasan committed:
Added interface

supermegaciaccount committed:
Added InputSplit method

zentol commented:
We ran into a problem when using the UnsplittableInput interface: the example stops working.
"Cannot compute input splits for Data Source: java.lang.NegativeArraySizeException" was the error message we got.

I was able to resolve it by overriding the createInputSplits method, with one that returns a list of a single GenericInputSplit().

rmetzger commented:
Hi @zentol;
I'm quite sure that this is a bug inside the system, so you don't have to do any workarounds. @StephanEwen can probably fix it within seconds ;)
I'll ask him tomorrow to fix it.

emrehasan committed:
Merge branch 'master' of https://github.com/emrehasan/stratosphere

StephanEwen commented:
Manually merged (because of conflicts and renaming) in 4d73ef004199fb933bf4669da63e4254ad115c3f

Thanks guys, that was good work!
